File: src/srbpy/alignment/pqx.py
import abc
from enum import Enum
import numpy as np
from skspatial.objects import Point
from .ext import *


class LeftRightEnum(Enum):
    NoneLR = 0
    Left = -1
    Right = 1


class EITypeID(Enum):
    NoType = 0
    Line = 1
    Arc = 2
    ZHY = 3
    YHZ = 4
    BigRtoSmallR = 5
    SmallRtoBigR = 6

    @staticmethod
    def from_value(v: int):
        for item in EITypeID:
            if item.value == v:
                return item
            else:
                continue
        raise Exception("无法获取枚举.")


class PQXElement(metaclass=abc.ABCMeta):
    """
    平曲线元素基类
    """
    typeid = EITypeID.NoType
    start_point = Point([0, 0])
    start_angle = Angle(deg=0)
    left_right = LeftRightEnum.NoneLR
    __length = 0
    __end_point = Point([0, 0])
    __end_angle = Angle(deg=0)

    def __init__(self, idd: EITypeID, pt: Point, st_angle: Angle, lr: LeftRightEnum):
        self.typeid = idd
        self.start_point = pt
        self.start_angle = st_angle
        self.left_right = lr

    @abc.abstractmethod
    def get_point_on_curve(self, l_from_st):
        pass

    @property
    @abc.abstractmethod
    def length(self):
        pass

    @property
    @abc.abstractmethod
    def end_angle(self):
        pass

    @property
    @abc.abstractmethod
    def end_point(self):
        pass


class Straight(PQXElement):
    __length = 0

    def __init__(self, length, st, sdir, idd=EITypeID.Line, lr=LeftRightEnum.NoneLR):
        super().__init__(idd, st, sdir, lr)
        self.__length = length

    def get_point_on_curve(self, length):
        x = length
        y = 0
        res = rotate2d(Vector([y, x]), self.start_angle * -1)
        return self.start_point + res
        pass

    @property
    def length(self):
        return self.__length

    @property
    def end_angle(self):
        return self.start_angle

    @property
    def end_point(self):
        return self.get_point_on_curve(self.__length)


class Arc(PQXElement):
    __length = 0
    __radius = 0

    def __init__(self, r, length, st, sdir, direnum, idd=EITypeID.Arc):
        super().__init__(idd, st, sdir, direnum)
        self.__length = length
        self.__radius = r

    @property
    def length(self):
        return self.__length

    @property
    def radius(self):
        return self.__radius

    @property
    def end_point(self):
        return self.get_point_on_curve(self.__length)

    @property
    def end_angle(self):
        return self.start_angle + Angle.from_rad(self.__length / self.__radius) * float(self.left_right.value)

    def get_point_on_curve(self, l_from_st):
        rad = l_from_st / self.__radius
        x = self.__radius * np.sin(rad)
        y = self.left_right.value * self.__radius * (1 - np.cos(rad))
        res = rotate2d(Vector([y, x]), self.start_angle * -1)
        return self.start_point + res


class Sprial(PQXElement):

    def __init__(self, idd: EITypeID, a: float, sr: float, er: float, st: Point, sa: Angle, lr_dir: LeftRightEnum):
        super().__init__(idd, st, sa, lr_dir)
        self.__a = a
        self.__start_radius = sr
        self.__end_radius = er

    def __ab__(self, value: float, i: int, j: int, k: int, ll: int):
        rho_a = 1.0 / self.__start_radius
        rho_b = 1.0 / self.__end_radius
        const_a = rho_a
        const_b = (rho_b - rho_a) / (2.0 * self.length)
        return np.power(const_a, i) * np.power(const_b, j) / float(k) * np.power(value, ll)

    @staticmethod
    def eval_xy(R_st, A, l, idd):
        if idd in [3, 5]:
            x = l - l ** 3 / (6 * R_st ** 2) - l ** 4 / (8 * A ** 2 * R_st) + l ** 6 / (72 * A ** 2 * R_st ** 3) + l ** 5 * (A ** 4 - 3 * R_st ** 4) / (
                    120 * A ** 4 * R_st ** 4) + l ** 7 * (-A ** 4 + 45 * R_st ** 4) / (5040 * A ** 4 * R_st ** 6) - l ** 10 / (
                        2880 * A ** 6 * R_st ** 3) + l ** 8 * (-A ** 4 + 5 * R_st ** 4) / (1920 * A ** 6 * R_st ** 5) - l ** 11 / (
                        8448 * A ** 8 * R_st ** 2) + l ** 9 * (-2 * A ** 4 + R_st ** 4) / (3456 * A ** 8 * R_st ** 4) - l ** 12 / (
                        46080 * A ** 10 * R_st) - l ** 13 / (599040 * A ** 12)
            y = l ** 2 / (2 * R_st) - l ** 4 / (24 * R_st ** 3) + l ** 3 / (6 * A ** 2) - l ** 5 / (20 * A ** 2 * R_st ** 2) + l ** 6 * (
                    A ** 4 - 15 * R_st ** 4) / (720 * A ** 4 * R_st ** 5) + l ** 8 * (-A ** 4 + 105 * R_st ** 4) / (40320 * A ** 4 * R_st ** 7) + l ** 7 * (
                        A ** 4 - R_st ** 4) / (336 * A ** 6 * R_st ** 4) + l ** 9 * (-A ** 4 + 15 * R_st ** 4) / (12960 * A ** 6 * R_st ** 6) - l ** 12 / (
                        27648 * A ** 8 * R_st ** 3) + l ** 10 * (-2 * A ** 4 + 5 * R_st ** 4) / (19200 * A ** 8 * R_st ** 5) - l ** 13 / (
                        99840 * A ** 10 * R_st ** 2) + l ** 11 * (-10 * A ** 4 + 3 * R_st ** 4) / (126720 * A ** 10 * R_st ** 4) - l ** 14 / (
                        645120 * A ** 12 * R_st) - l ** 15 / (9676800 * A ** 14)
        elif idd in [4, 6]:
            x = l - l ** 3 / (6 * R_st ** 2) + l ** 4 / (8 * A ** 2 * R_st) - l ** 6 / (72 * A ** 2 * R_st ** 3) + l ** 5 * (A ** 4 - 3 * R_st ** 4) / (
                    120 * A ** 4 * R_st ** 4) + l ** 7 * (-A ** 4 + 45 * R_st ** 4) / (5040 * A ** 4 * R_st ** 6) + l ** 10 / (
                        2880 * A ** 6 * R_st ** 3) + l ** 8 * (A ** 4 - 5 * R_st ** 4) / (1920 * A ** 6 * R_st ** 5) - l ** 11 / (
                        8448 * A ** 8 * R_st ** 2) + l ** 9 * (-2 * A ** 4 + R_st ** 4) / (3456 * A ** 8 * R_st ** 4) + l ** 12 / (
                        46080 * A ** 10 * R_st) - l ** 13 / (599040 * A ** 12)
            y = l ** 2 / (2 * R_st) - l ** 4 / (24 * R_st ** 3) - l ** 3 / (6 * A ** 2) + l ** 5 / (20 * A ** 2 * R_st ** 2) + l ** 6 * (
                    A ** 4 - 15 * R_st ** 4) / (720 * A ** 4 * R_st ** 5) + l ** 8 * (-A ** 4 + 105 * R_st ** 4) / (40320 * A ** 4 * R_st ** 7) + l ** 7 * (
                        -A ** 4 + R_st ** 4) / (336 * A ** 6 * R_st ** 4) + l ** 9 * (A ** 4 - 15 * R_st ** 4) / (12960 * A ** 6 * R_st ** 6) - l ** 12 / (
                        27648 * A ** 8 * R_st ** 3) + l ** 10 * (-2 * A ** 4 + 5 * R_st ** 4) / (19200 * A ** 8 * R_st ** 5) + l ** 13 / (
                        99840 * A ** 10 * R_st ** 2) + l ** 11 * (10 * A ** 4 - 3 * R_st ** 4) / (126720 * A ** 10 * R_st ** 4) - l ** 14 / (
                        645120 * A ** 12 * R_st) + l ** 15 / (9676800 * A ** 14)
        else:
            raise Exception("缓和曲线类型错误.")
        return x, y

    def get_point_on_curve_old(self, l_from_st):
        x = l_from_st - self.__ab__(l_from_st, 2, 0, 6, 3) - self.__ab__(l_from_st, 1, 1, 4, 4) \
            - self.__ab__(l_from_st, 0, 2, 10, 5) + self.__ab__(l_from_st, 4, 0, 120, 5) + self.__ab__(l_from_st, 3, 1, 36, 6) \
            + self.__ab__(l_from_st, 2, 2, 28, 7) + self.__ab__(l_from_st, 1, 3, 48, 8) + self.__ab__(l_from_st, 0, 4, 216, 9)

        y = self.__ab__(l_from_st, 1, 0, 2, 2) + self.__ab__(l_from_st, 0, 1, 3, 3) - self.__ab__(l_from_st, 3, 0, 24, 4) \
            - self.__ab__(l_from_st, 2, 1, 10, 5) - self.__ab__(l_from_st, 1, 2, 12, 6) + self.__ab__(l_from_st, 5, 0, 720, 6) \
            - self.__ab__(l_from_st, 0, 3, 42, 7) + self.__ab__(l_from_st, 4, 1, 168, 7) + self.__ab__(l_from_st, 3, 2, 96, 8) \
            + self.__ab__(l_from_st, 2, 3, 108, 9)

        y *= self.left_right.value
        res = rotate2d(Vector([y, x]), self.start_angle * -1)
        return self.start_point + res

    def get_point_on_curve(self, l_from_st):
        x0, y0 = self.eval_xy(self.__start_radius, self.__a, l_from_st, self.typeid.value)
        y0 *= self.left_right.value
        res = rotate2d(Vector([float(y0), float(x0)]), self.start_angle * -1)
        return self.start_point + res

    @property
    def length(self):
        return np.abs(self.__a * self.__a / self.__start_radius - self.__a * self.__a / self.__end_radius)

    @property
    def end_angle(self):
        ang = Angle.from_rad(0.5 * self.__a ** 2 * np.abs(1.0 / (self.__start_radius ** 2) - 1.0 / (self.__end_radius ** 2)))
        return self.start_angle + ang * float(self.left_right.value)

    @property
    def end_point(self):
        return self.get_point_on_curve(self.length)


class PQX(object):
    start_pk = 0
    start_point = Point([0, 0])
    start_angle = Angle(deg=0)

    def __init__(self, file=""):
        cur_point = Point([0, 0])
        cur_angle = Angle(deg=0.0)
        elem_collection = []
        with open(file, 'r') as fid:
            text = fid.readlines()
        for i, line in enumerate(text):
            if i == 0:
                self.start_pk = float(line)
            elif i == 1:
                xx = line.split(',')
                start_x = float(xx[0])
                start_y = float(xx[1])
                start_ang_in_rad = float(xx[2])
                self.start_point = Point([start_y, start_x])
                self.start_angle = Angle(deg=start_ang_in_rad / np.pi * 180.0)
                cur_point = self.start_point
                cur_angle = self.start_angle
            else:
                xx = line.split(',')
                if line.startswith("//"):
                    continue
                elif len(xx) == 3 and int(xx[2]) == 0:
                    break
                else:
                    item = None
                    try:
                        type_id = int(xx[0])
                    except Exception:
                        raise Exception("无法解析元素1.")

                    if type_id == 1:  # 直线，长度，方位角
                        ll = float(xx[1])
                        item = Straight(ll, cur_point, cur_angle)
                    elif type_id == 2:
                        rr = float(xx[1])
                        ll = float(xx[2])
                        s_dir = LeftRightEnum.Left if int(xx[3]) == -1 else LeftRightEnum.Right
                        item = Arc(rr, ll, cur_point, cur_angle, s_dir)
                    elif type_id == 3:  #
                        idd = EITypeID.from_value(type_id)
                        aa = float(xx[1])
                        end_r = float(xx[2])
                        lr_dir = LeftRightEnum.Left if int(xx[3]) == -1 else LeftRightEnum.Right
                        item = Sprial(idd, aa, 1e39, end_r, cur_point, cur_angle, lr_dir)
                    elif type_id == 4:
                        idd = EITypeID.from_value(type_id)
                        aa = float(xx[1])
                        st_r = float(xx[2])
                        lr_dir = LeftRightEnum.Left if int(xx[3]) == -1 else LeftRightEnum.Right
                        item = Sprial(idd, aa, st_r, 1e39, cur_point, cur_angle, lr_dir)
                    elif type_id == 5 or type_id == 6:
                        idd = EITypeID.from_value(type_id)
                        aa = float(xx[1])
                        st_r = float(xx[2])
                        end_r = float(xx[3])
                        lr_dir = LeftRightEnum.Left if int(xx[4]) == -1 else LeftRightEnum.Right
                        item = Sprial(idd, aa, st_r, end_r, cur_point, cur_angle, lr_dir)
                    else:
                        raise Exception("读取ICD文件错误.")
                    elem_collection.append(item)
                    cur_angle = elem_collection[-1].end_angle
                    cur_point = elem_collection[-1].end_point
        self.__elem_collection = elem_collection

    @property
    def end_pk(self):
        if len(self.__elem_collection) == 0:
            return self.start_pk
        else:
            return self.start_pk + sum(item.length for item in self.__elem_collection)

    def get_dir(self, pk: float, delta: float = 0.00001):
        """
        获取任意里程切线单位向量
        Parameters
        ----------
        pk : float
            任意里程
        delta : float , optional
            精度值 , 默认0.00001

        Returns
        -------
        out : list
            切线单位向量坐标值
        """
        x0 = self.get_coordinate(pk - delta)[0]
        y0 = self.get_coordinate(pk - delta)[1]
        x1 = self.get_coordinate(pk + delta)[0]
        y1 = self.get_coordinate(pk + delta)[1]
        ll = np.sqrt((x0 - x1) * (x0 - x1) + (y0 - y1) * (y0 - y1))
        return [(x1 - x0) / ll, (y1 - y0) / ll]

    def get_coordinate(self, pk: float):
        """
        返回任意里程坐标.
        Parameters
        ----------
        pk : float
            里程桩号.
        Returns
        -------
        out : list
            大地坐标
        """
        if pk < self.start_pk:
            return self.get_coordinate(self.start_pk)
        elif pk <= self.end_pk:
            tmp = [item.length for item in self.__elem_collection]
            len_sum_up = [self.start_pk]
            for ll in tmp:
                len_sum_up.append(len_sum_up[-1] + ll)
            aa = 0
            for aa, val in enumerate(len_sum_up):
                if val > pk:
                    break
            aa = aa - 1

            if pk == self.end_pk:
                pre_pk = len_sum_up[-2]
                # nex_pk = len_sum_up[-1]
            else:
                pre_pk = len_sum_up[aa]
                # nex_pk = len_sum_up[aa + 1]
            ll = pk - pre_pk
            res = self.__elem_collection[aa].get_point_on_curve(ll)
            return res[0], res[1]

    def get_station_by_point(self, x0: float, y0: float, step: int = 20, delta: float = 1e-9) -> float:
        """
        根据单点获得正交桩号(可能多解)
        Parameters
        ----------
        x0 : float
            坐标X
        y0 : float
            坐标Y
        step : int
            步长，越大精度越高，但求解时间增加
        delta : float
            精度

        Returns
        -------
        out : float
            返回对应桩号
        """
        res = []
        dk = (self.end_pk - self.start_pk) / step
        target = Point([x0, y0])
        for i in range(step):
            if i == 17:
                test = True
            k1 = self.start_pk + dk * i
            k2 = k1 + dk
            k0, err = self.__binary_test__(k1, k2, target, delta)
            res.append([k0, err])
        res.sort(key=lambda x: x[1])
        return res[0][0]

    def get_side(self, x0: float, y0: float) -> int:
        pk_center = self.get_station_by_point(x0, y0)
        center = Point(list(self.get_coordinate(pk_center)))
        v_dir = Vector(list(self.get_dir(pk_center)))
        v_cross = Point([x0, y0]) - center
        if length(v_cross) == 0:
            return 0
        else:
            ang = Angle.from_rad(v_dir.angle_signed(v_cross))
            if ang.to_degrees() < 180:
                return -1
            else:
                return 1

    def __solve_closer__(self, pk1: float, pk2: float, pt: Point):
        """
        二分法
        Parameters
        ----------
        pk1 : float
            里程起点
        pk2 : float
            里程终点
        pt : Point
            逼近点

        Returns
        -------
        out : float , float , float
            迭代下限,上限,评价值
        """
        assert pk1 <= pk2
        p1 = Point(list(self.get_coordinate(pk1)))
        p2 = Point(list(self.get_coordinate(pk2)))
        if p1.distance_point(pt) < p2.distance_point(pt):
            return pk1, 0.5 * (pk1 + pk2), p1.distance_point(pt)
        else:
            return 0.5 * (pk1 + pk2), pk2, p2.distance_point(pt)

    def __binary_test__(self, pk1: float, pk2: float, pt: Point, delta: float) -> tuple:
        """

        Parameters
        ----------
        pk1
        pk2
        pt
        delta

        Returns
        -------
        out : (r,dist)
            最优解,误差
        """
        mid = 0
        while np.abs(pk1 - pk2) > delta:
            pk1, pk2, dist = self.__solve_closer__(pk1, pk2, pt)
            mid = 0.5 * (pk1 + pk2)
        res1 = mid
        pk1, pk2, dist = self.__solve_closer__(res1, res1, pt)
        return res1, dist

File: src/srbpy/alignment/PQX.cpp
#include <algorithm>
#include <cassert>
#include <sstream>
#include <stdexcept>
#include <string>
#include "PQX.h"
#include "PQXElement.h"
#include "Straight.h"
#include "Arc.h"
#include "Sprial.h"

using namespace std;

// 替代 boost::split(..., is_any_of(","), token_compress_on)：
// 按分隔符切分 UTF-8 字符串，合并相邻分隔符并丢弃空 token。
// ICD 数值字段和分隔符均为 ASCII，可直接按字节解析。
static vector<string> split_str(const string &input, char delim) {
    vector<string> result;
    string token;
    for (char ch : input) {
        if (ch == delim) {
            if (!token.empty()) {
                result.push_back(token);
                token.clear();
            }
        } else {
            token.push_back(ch);
        }
    }
    if (!token.empty()) {
        result.push_back(token);
    }
    return result;
}

PQX::PQX(void) {
    start_pk = 0;
    start_point = Vector();
    start_angle = Angle();
    end_pk = 0;
    __rebuild_index();
}

PQX::PQX(const std::filesystem::path &filepath) {
    start_pk = 0;
    start_point = Vector();
    start_angle = Angle();
    Vector cur_point = Vector(0, 0);
    Angle cur_angle = Angle(Degree(0));
    string t1;
    vector<string> text;

    ifstream fid(filepath, ios::in | ios::binary);
    if (!fid.is_open()) {
        throw runtime_error("无法打开 ICD 文件: " + filepath.generic_u8string());
    }

    while (getline(fid, t1)) {
        if (!t1.empty() && t1.back() == '\r') {
            t1.pop_back();
        }
        text.push_back(t1);
    }
    fid.close();
    stringstream ss;
    for (size_t i = 0; i < text.size(); ++i) {
        if (i != 0)
            ss << "\n";
        ss << text[i];
    }
    PQX::ICDText = ss.str();

    string line;

    for (size_t i = 0; i < text.size(); i++) {
        line = text[i];

        if (i == 0) {
            start_pk = stod(line);
        } else if (i == 1) {
            vector<string> xx = split_str(line, ',');
            double start_x = stod(xx[0]);
            double start_y = stod(xx[1]);
            double start_ang_in_rad = stod(xx[2]);
            start_point = Vector(start_y, start_x);
            start_angle = Angle(Degree(start_ang_in_rad / Angle::PI * 180.0));
            cur_point = start_point;
            cur_angle = start_angle;
        } else {
            vector<string> xx = split_str(line, ',');
            if (line.rfind("//", 0) == 0) {
                continue;
            } else if (xx.size() == 3 && stoi(xx[2]) == 0) {
                break;
            } else {
                PQXElement *item = nullptr;
                Straight str;
                Sprial spr;
                Arc arc;
                int type_id = stoi(xx[0]);
                double aa, ll, st_r, end_r, rr;
                LeftRightEnum lr_dir;
                EITypeID idd;
                switch (type_id) {
                    case 1:
                        ll = stod(xx[1]);
                        // 补充考虑突变夹角
                        if (xx.size() == 3) {
                            cur_angle = stod(xx[2]);
                        }
                        item = new Straight(ll, cur_point, cur_angle);
                        break;
                    case 2:
                        rr = stod(xx[1]);
                        ll = stod(xx[2]);
                        lr_dir = LeftRightEnum(stoi(xx[3]));
                        // 补充考虑突变夹角
                        if (xx.size() == 5) {
                            cur_angle = stod(xx[4]);
                        }
                        item = new Arc(rr, ll, cur_point, cur_angle, lr_dir);
                        break;
                    case 3:
                        idd = EITypeID(type_id);
                        aa = stod(xx[1]);
                        end_r = stod(xx[2]);
                        lr_dir = LeftRightEnum(stoi(xx[3]));
                        item = new Sprial(idd, aa, 1e39, end_r, cur_point, cur_angle, lr_dir);
                        break;
                    case 4:
                        idd = EITypeID(type_id);
                        aa = stod(xx[1]);
                        st_r = stod(xx[2]);
                        lr_dir = LeftRightEnum(stoi(xx[3]));
                        item = new Sprial(idd, aa, st_r, 1e39, cur_point, cur_angle, lr_dir);
                        break;
                    case 5:
                    case 6:
                        idd = EITypeID(type_id);
                        aa = stod(xx[1]);
                        st_r = stod(xx[2]);
                        end_r = stod(xx[3]);
                        lr_dir = LeftRightEnum(stoi(xx[4]));
                        item = new Sprial(idd, aa, st_r, end_r, cur_point, cur_angle, lr_dir);
                        // 怀疑构造函数的终点半径输入错误；
                        break;
                    default:
                        throw runtime_error("读取 ICD 文件错误：未知的路线元素类型。");
                }

                elem_collection.push_back(item);
                cur_angle = elem_collection.back()->end_angle();
                cur_point = elem_collection.back()->end_point();;
            }


        }

    }

    end_pk = __get_end_pk();
    __rebuild_index();
}

void PQX::__rebuild_index() {
    __len_sum_up.clear();
    __len_sum_up.push_back(start_pk);
    for (auto l : elem_collection) {
        __len_sum_up.push_back(__len_sum_up.back() + l->length());
    }
}

// 定位 pk 落在哪个单元，返回单元下标 idx 与该单元内局部弧长 local_l。
// 调用前应保证 elem_collection 非空、且 start_pk <= pk <= end_pk。
void PQX::locate(double pk, int &idx, double &local_l) const {
    int n = (int) elem_collection.size();
    int aa = 0;
    for (aa = 0; aa < (int) __len_sum_up.size(); aa++) {
        if (__len_sum_up[aa] > pk) {
            break;
        }
    }
    if (aa == (int) __len_sum_up.size()) {
        aa = aa - 1;
    }
    aa = aa - 1;
    double pre_pk;
    if (pk >= end_pk) {
        // 末端归一：固定落在最后一个单元的终点
        aa = n - 1;
        pre_pk = __len_sum_up[__len_sum_up.size() - 2];
    } else {
        if (aa < 0) aa = 0;
        pre_pk = __len_sum_up[aa];
    }
    idx = aa;
    local_l = pk - pre_pk;
}

double PQX::__get_end_pk() const {
    if (elem_collection.size() == 0) {
        return start_pk;
    } else {
        double totalLength = 0;
        for (auto l :elem_collection) {
            totalLength = totalLength + l->length();
        }
        return start_pk + totalLength;
    }
}

Vector PQX::get_coordinate(double pk) const {
    if (elem_collection.empty()) {
        return start_point;
    }
    if (pk < start_pk) {
        return get_coordinate(start_pk);
    } else if (pk <= end_pk) {
        int idx;
        double local_l;
        locate(pk, idx, local_l);
        return elem_collection[idx]->get_point_on_curve(local_l);
    }
    return Vector();
}

Vector PQX::get_dir(double pk, double delta) const {
    // 解析切线：直接取该点切线方位角，避免对大坐标做数值差分造成的抵消误差。
    // delta 参数已废弃，仅为保持接口兼容而保留。
    (void) delta;
    if (elem_collection.empty()) {
        return Vector(start_angle.Sin(), start_angle.Cos());
    }
    double cpk = pk;
    if (cpk < start_pk) cpk = start_pk;
    if (cpk > end_pk) cpk = end_pk;
    int idx;
    double local_l;
    locate(cpk, idx, local_l);
    Angle phi = elem_collection[idx]->dir_angle_on_curve(local_l);
    // 与 get_point_on_curve 的坐标约定一致：单位切线 = (sin φ, cos φ)
    return Vector(phi.Sin(), phi.Cos());
}

double PQX::get_station_by_point(double x0, double y0, int step, double delta) {
    vector<pair<double, double>> res;
    double k0err[2];
    double dk = (end_pk - start_pk) / step;
    Vector target = Vector(x0, y0);
    double k1, k2;
    for (int i = 0; i < step; i++) {
        k1 = start_pk + dk * i;
        k2 = k1 + dk;
        __binary_test__(k1, k2, target, delta, k0err);
        res.push_back(pair<double, double>{k0err[0], k0err[1]});
    }
    sort(res.begin(), res.end(),
         [](pair<double, double> a, pair<double, double> b) -> bool { return (a.second < b.second); });

    return res[0].first;
}


double PQX::get_station_by_point2(double x0, double y0, double x1, double y1, double delta) {
    Vector pt0 = Vector(x0, y0);
    Vector pt1 = Vector(x1, y1);
    double ret[2] = {start_pk, end_pk};
    while (ret[1] - ret[0] > delta) {
        __solve_coincidence(pt0, pt1, ret);
    }
    return (ret[0] + ret[1]) * 0.5;
}


int PQX::get_side(double x0, double y0) {
    double pk_center = get_station_by_point(x0, y0);
    Vector center = get_coordinate(pk_center);
    Vector v_dir = get_dir(pk_center);
    Vector v_cross = Vector(x0, y0) - center;
    if (v_cross.length() == 0) {
        return 0;
    } else {
        Angle ang = Angle(v_dir.angle_signed(v_cross));
        if (ang.GetDegree() < 0) {
            return 1;
        } else {
            return -1;
        }
    }
    return 0;
}

void PQX::__solve_closer(const Vector &pt, double ret[3]) {
    double pk1 = ret[0];
    double pk2 = ret[1];
    assert(pk1 <= pk2);
    Vector p1 = get_coordinate(pk1);
    Vector p2;
    // 修改了因pk2超过endpk造成的错误 -bill 10.15
    // Vector p2 = get_coordinate(pk2);
    if (abs(pk2 - end_pk) < 1.0e-6) {
        p2 = get_coordinate(end_pk);
    } else {
        p2 = get_coordinate(pk2);
    }
    if (p1.distance_point(pt) < p2.distance_point(pt)) {
        ret[0] = pk1;
        ret[1] = 0.5 * (pk1 + pk2);
        ret[2] = p1.distance_point(pt);
    } else {
        ret[0] = 0.5 * (pk1 + pk2);
        ret[1] = pk2;
        ret[2] = p2.distance_point(pt);
    }
}


/// 二分求解
/// \param pk1
/// \param pk2
/// \param pt
/// \param delta
/// \param ret 返回值分别为最优解及误差.
void PQX::__binary_test__(double pk1, double pk2, const Vector &pt, double delta, double ret[2]) {
    double mid = 0;
    double inputs[3] = {pk1, pk2, 0};
    while (abs(inputs[1] - inputs[0]) > delta) {
        __solve_closer(pt, inputs);
        mid = 0.5 * (inputs[0] + inputs[1]);
    }
    double res1 = mid;
    inputs[0] = res1;
    inputs[1] = res1;
    __solve_closer(pt, inputs);
    ret[0] = res1;
    ret[1] = inputs[2];
}


/// 求解与pt0,pt1 共线的点.
/// \param pt0
/// \param pt1
/// \param ret [k0,k1]
void PQX::__solve_coincidence(const Vector &pt0, const Vector &pt1, double *ret) {
    Vector dir = pt1 - pt0;
    assert(ret[1] > ret[0]);
    double km=(ret[0] + ret[1]) * 0.5;
    Vector v0 = get_coordinate(ret[0]) - pt0;
    Vector v1 = get_coordinate(ret[1]) - pt0;
    Vector vm = get_coordinate(km) - pt0;
    double ang0 = dir.angle_signed(v0);
    double ang1 = dir.angle_signed(v1);
    double ang_m = dir.angle_signed(vm);
    if (ang0 * ang1 > 0) {
        // 同向，反射
        double dl = ret[1] - ret[0];
        if (abs(ang0) < abs(ang1)) {
            ret[1] = ret[0];
            ret[0] = ret[0] - dl;
        } else {
            ret[0] = ret[1];
            ret[1] = ret[1] + dl;
        }

    } else {
        // 不同向，二分
        if (ang0 * ang_m > 0) {
            ret[0] = km;
        } else {
            ret[1] = km;
        }
    }
}

